File: World Finals 2009/C.cpp
// Problem: Doubly-sorted Grid
// Language: C++
// Author: KirarinSnow
// Usage: g++ thisfile.cpp -o executable && ./executable <input.in >output.out


#include <stdio.h>
#include <iostream>
#include <string.h>

using namespace std;

const int MOD = 10007;

int d[1<<20][26][11];
bool b[1<<20][26][11];
char g[10][10];

int row, col;


int dp(int p, int c, int k)
{
    if (b[p][c][k] == false)
    {
	int t = 0;
	if (p == (((1<<row)-1)<<col))
	{
	    t = 1;
	}
	else
	{
	    if (k < 0)
	    {
		if (c != 0)
		{
		    t += dp(p, c-1, col-1);
		}
	    }
	    else
	    {	
		t += dp(p, c, k-1);
		
		int pp = p;
		int x = col;
		int y = 0;
		int z = 0;
		while (x > k)
		{
		    if (pp%2 == 0)
		    {
			x--;
		    }
		    else
		    {
			z++;
		    }
		    y++;
		    pp /= 2;
		}
		
		
		if ((y >= 2) && (((p%(1<<y))>>(y-2)) == 1))
		{
		    char ch = g[z-1][k];
		    if (ch == '.' || ch == (char)((int)'a'+c))
		    {
			int flip = (pp << y) + (2 << (y-2)) + (p%(1<<(y-2)));
			t += dp(flip, c, k);
		    }
		}
		
	    }
	}
	d[p][c][k] = t%MOD;
	b[p][c][k] = true;
    }
    return d[p][c][k];
}

int solve()
{
    memset(d, 0, sizeof d);
    memset(b, false, sizeof b);
    memset(g, 0, sizeof g);

    cin >> row >> col;

    for (int i = 0; i < row; i++)
    {
	for (int j = 0; j < col; j++)
	{
	    cin >> g[i][j];
	}
    }

    return dp((1<<row)-1, 25, col-1);
}

int main()
{
    int cases;
    cin >> cases;
    for (int i = 1; i <= cases; i++)
    {
	printf("Case #%d: %d\n", i, solve());
    }
    
    return 0;
}

File: World Finals 2009/C.py
#!/usr/bin/python
#
# Problem: Doubly-sorted Grid
# Language: Python
# Author: KirarinSnow
# Usage: python thisfile.py <input.in >output.out
# Comments: Times out on large.


MOD = 10007

def compute():
    def proc(p, k):
        if (p, k) not in pd:
            pp = p
            x = C
            y = 0
            z = -1
            while x > k:
                if pp%2 == 0:
                    x -= 1
                else:
                    z += 1
                y += 1
                pp /= 2
            infl = (y >= 2 and ((p%(1<<y))>>(y-2)) == 1)
            flip = (pp << y) + (2 << (y-2)) + (p%(1<<(y-2))) if infl else None
            pd[(p, k)] = (infl, z, flip)
        return pd[(p, k)]


    def dp(p, c, k):
        if (p, c, k) not in d:
            t = 0
            if p == ((1<<R)-1)<<C:
                t = 1
            else:
                if k == -1:
                    if c != ord('a'):
                        t += dp(p, c-1, C-1)
                        
                else:
                    t += dp(p, c, k-1)
                    infl, row, flip = proc(p, k)
                    
                    if infl:
                        ch = g[row][k]
                        if ch == '.' or ch == chr(c):
                            t += dp(flip, c, k)
            d[(p, c, k)] = t%MOD
        return d[(p, c, k)]

    d = {}
    pd = {}
    R, C = map(int, raw_input().split())
    g = [raw_input() for j in range(R)]

    return dp((1<<R)-1, ord('z'), C-1)
 
for i in range(int(raw_input())):
    print "Case #%d: %d" % (i+1, compute())
